blog-content/interesting-sessions.md

```diff
@@ -0,0 +1,75 @@
+
+- If the part contains a number, treat the part as an ID.
+- Split the part on capitals or separator characters (’-’, ‘_’, ‘~’, ‘+’). If any of these doesn’t contain a vowel or contains more than 5 sequential consonants, treat the part as an ID.
+
+There are a few cases where this heuristic won’t work. IDs generated from dictionary words (e.g. “diamond-hamburger-board”) will be treated as separate pages. And any static pages with numbers or strings of consonants (e.g. in acronyms) will be treated as IDs. Both of these situations won’t have a huge impact for most apps, but will add some noise to the results. A better approach may look back at previous journeys to merge similar URLs, but this can have issues too, e.g. for handling new pages.
+
+After applying these normalization steps, we can enumerate all of the state transitions in a user journey. For each step in the journey, we can calculate the bigram probability of the next url given the current url using all of the other sessions we’ve seen, creating a Markov chain. An interesting session should be one where the state transitions are less probable. For a Highlight user session, an example journey with its corresponding probabilities looks like this:
+
+| page | next_page | probability |
+| --- | --- | --- |
+| START | /\{id-1\}/sessions/\{id-2\} | 0.22617124394184200000 |
+| /\{id-1\}/sessions/\{id-2\} | /\{id-1\}/errors/\{id-2\} | 0.02150537634408600000 |
+| /\{id-1\}/errors/\{id-2\} | /\{id-1\}/errors/\{id-2\}/instances/\{id-3\} | 0.05154639175257730000 |
+| /\{id-1\}/errors/\{id-2\}/instances/\{id-3\} | /\{id-1\}/sessions | 0.06060606060606060000 |
+| /\{id-1\}/sessions | /\{id-1\}/sessions/\{id-2\} | 0.47959183673469400000 |
+| /\{id-1\}/sessions/\{id-2\} | END | 0.68817204301075300000 |
+
+We can calculate the probability of the entire session by multiplying the probabilities together. In the above example, we get a total probability of `5.01x10^-6`. This is the probability of that exact user journey happening, assuming a page was randomly chosen based only on the current page. We could calculate this probability for all sessions, then find the sessions with the lowest probability. This has a couple of issues though:
+
+1. Longer sessions will be favored, because probabilities are multiplied together and the probability of every step is less than or equal 1.
+2. Sessions with state transitions in parts of the app with a high branching factor (where many possible pages can be reached from the current page) will be favored, as the expected probability is lower. For example, if there are five pages a user can visit from a current page, each with equal probability, the step’s probability will be `.2` regardless of which page is chosen (and in this example, each page is equally “interesting”, so the transition shouldn’t affect the overall “interestingness”).
```

Comment:
this is a bit confusing, wonder if there is a simpler way to explain this. 

chatgpt proposal:
> In parts of the app where you can go to a lot of different pages from the page you're on, sessions that move around a lot will be more likely. Think of it this way: if you're on a page that can take you to five different pages, and you have an equal chance of going to any of these five pages, the chance you'll go to any one page is 20% (or .2). It doesn't matter which page you choose to go to next because each page is just as "interesting" as the others, so moving from one page to another doesn't change how "interesting" your session is overall.

Reply:
cleaned this up a bit